fabric/src/main/java/io/github/retrooper/packetevents/handler/PacketEncoder.java

```diff
@@ -46,15 +59,100 @@ public void write(ChannelHandlerContext ctx, Object msg, ChannelPromise promise)
             ctx.write(msg, promise);
             return;
         }
+
+        // Handle promise management first (matches Spigot)
+        ChannelPromise oldPromise = this.promise != null && !this.promise.isSuccess() ? this.promise : null;
+        promise.addListener(p -> this.promise = oldPromise);
+        this.promise = promise;
+
+        // Process the packet and execute post-send tasks (matches Spigot)
+        handlePacket(ctx, in, promise);
+
+        // Check for empty packets last (matches Spigot)
         if (!in.isReadable()) {
             in.release();
+            throw CancelPacketException.INSTANCE;
```

Comment:
as this isn't a `MessageToMessageEncoder` (like spigot), just do `return` instead of throwing an exception

Reply:
which part are you uncertain about?